### DISCUSSION TITLE
Wie genau einrichten?

### DISCUSSION THREAD

**Discussion by gamer123** [spali/go-rscp#16] [Q&A] [answered]
Hi @spali,

danke für die Implementierung der RCSP Schnittstelle. 

Ich habe versucht diese in Betrieb zu nehmen(mit HA):
Die Beschreibung ist gut, jedoch stellt sich mir die Frage welchen Benutzer und welcher Key bzw wie ich diesen ermittle und wie dieser hier eintragen werden muss. 
Ich wollte die Lokale Version verwenden. 
D.h. 

host <ip>
user <username>
password <password>
key <aes key>

host/IP ist klar
Port ist nicht einstellbar daher gehe ich davon aus das es der standard Port am E3DC ist (glaube 5033). 
USER: local.user ist das korrekt ? 
PW: test123  ( am E3DC eingestellt)
KEY ? Was muss hier eingestellt werden ? 

Des Weiteren bringt Homeassistant einen Fehler in der e3dc.sh Datei. Der nicht näher spezifiziert wird. 
Vermutlich weil die Parameter nicht passen. 

Um Werte auszulesen reicht eine Implementierung bis zum Anfang des Schritts :
"Example script to execute something on E3DC"

Vielen Dank für eine evtl. Antwort/Hilfe

Julian

**Comment by spali** [comment]
> wollte die Lokale Version verwenden.

PS: diese Implementation geht nur lokal 😉 

> Port ist nicht einstellbar daher gehe ich davon aus das es der standard Port am E3DC ist (glaube 5033).

Ja default Port ist 5033, musst du nichts einstellen.

> USER: local.user ist das korrekt ?
> PW: test123 ( am E3DC eingestellt)

Gleiches Login wie in dem Web-Gui. User sollte im Normalfall eine Mail-Adresse sein.

> KEY ? Was muss hier eingestellt werden ?

Das ist ein zusätzliches "Passwort" welches du im Web-Gui setzen kannst. Zu finden unter "Personalisieren" -> "Benutzerprofil" -> "RSCP Passwort".

> Des Weiteren bringt Homeassistant einen Fehler in der e3dc.sh Datei. Der nicht näher spezifiziert wird. Vermutlich weil die Parameter nicht passen.

Würde ich auch sagen.

> Um Werte auszulesen reicht eine Implementierung bis zum Anfang des Schritts : "Example script to execute something on E3DC"

Korrekt.

**Comment by gamer123** [comment]
Hi 

danke für die schnelle Antwort. 

Leider bleibt der Fehler bestehen. 
Ich habe beides in der .config probiert  einmal mit und ohne <>
host 192.168.0.100
user mailadresse@web.de
password Test123!
key TEST123

leider weiterhin der Fehler
Command failed: /config/packages/e3dc/e3dc.sh

Die Datei sieht so aus 

`#!/bin/sh
set -e
export PATH=/config/bin:$PATH
cd "$(dirname $0)"

e3dc -file ./e3dc.json | jq -cM '{ "result": . }'`

Zusätzlich habe ich nochmal ein git clone der 0.1.0 Version in den bin Ordner gemacht, davor war die 0.0.8. 
curl -L https://github.com/spali/go-rscp/releases/download/v0.1.0/e3dc_0.1.0_Linux_x86_64.tar.gz | tar -xzf - e3dc

Muss der Host neu gestartet werden oder reicht ein Neustart von Homeassistant. 
By the way ich nutze Homeasistant OS auf einem eigenen Rechner.

**Reply by spali** [reply]
> Leider bleibt der Fehler bestehen. Ich habe beides in der .config probiert einmal mit und ohne <> host 192.168.0.100 user [mailadresse@web.de](mailto:mailadresse@web.de) password Test123! key TEST123

Muss ohne `<`, `>` sein.
Ich gehe davon aus, dass das mailto Zeugs vom Copy&Paste ist... das File sollte in etwa so aussehen:
```
host 192.168.0.100
user mailadresse@web.de
password Test123!
key TEST123
```
Key ist das gesetzte RSCP Passwort. Achtung für Key und Passwort gilt Gross-/Kleinschreibung beachten.

> Zusätzlich habe ich nochmal ein git clone der 0.1.0 Version in den bin Ordner gemacht, davor war die 0.0.8. curl -L https://github.com/spali/go-rscp/releases/download/v0.1.0/e3dc_0.1.0_Linux_x86_64.tar.gz | tar -xzf - e3dc

Ich habe die URL im Wiki angepasst 😉 
Ich gehe davon aus, dass du auch darauf geachtet hast, dass Linux und x86_64 die richtige Platform ist, welche du auch verwendest.

> Muss der Host neu gestartet werden oder reicht ein Neustart von Homeassistant. By the way ich nutze Homeasistant OS auf einem eigenen Rechner.

Home Assistant muss du nur neu starten, wenn du etwas an der Home Assistant config änderst (yaml Dateien). Änderungen am Skript oder ersetzen des Downloads, sollte sofort wirken.

> leider weiterhin der Fehler Command failed: /config/packages/e3dc/e3dc.sh
Leider zeigt Home Assistant nicht viel dazu an im eigenen Log. Hast du das Skript mal von Hand auf der Shell ausgeführt? 
Wenn ja bitte poste mal die Ausgabe (wenn möglich formatiert als Code der Übersichtshalber).
Was du auch direkt mal überprüfen kannst, ob das Skript executable Berechtigung hat.

**Reply by gamer123** [reply]
Danke nochmals für die schnelle Antwort. Ich glaube es liegt am letzten Punkt. Ich habe dem Skript keine Berechtigung erteilt. Wie würde ich das denn genau machen ? 

chmod +x /config/packages/e3dc/e3dc.sh ? 
Oder braucht das file im /bin/3edc/ auch Berechtigungen?

**Comment by spali** [comment, marked as answer]
Ja, das Skript und das heruntergeladene e3dc braucht die entsprechende Berechtigung. Mit dem Beispiel Download Befehl im Wiki sollte das normalerweise für das e3dc direkt der Fall sein. Notfalls kannst du alles mit folgenden Kommandos setzen:
```
chmod +x /config/packages/e3dc/e3dc.sh
chmod +x /config/bin/e3dc
```

**Comment by gamer123** [comment]
Hi @spali 

ok chmod war das Problem bez. dem Ausführen. 

Wenn ich das Skript von Hand in der Shell ausführe passiert folgendes.
![Shellgo RSCP](https://user-images.githubusercontent.com/5868875/167099277-3cdf6630-3903-45d3-a367-8ad28e9bfda7.jpg)
![Shellgo RSCP_pfad](https://user-images.githubusercontent.com/5868875/167099523-5805b394-5044-433c-9c11-70a13c818db8.jpg)
![Shellgo RSCP_e3dc](https://user-images.githubusercontent.com/5868875/167099534-b767086b-37ce-4af4-98bb-9ed2864da25d.jpg)


Es sieht so aus wie wenn das Skript nicht nicht ausführt oder die Datei wird nicht gefunden oder ?  

Danke

**Reply by gamer123** [reply]
Ich habe beim Skript ausführen ./e3dc vergessen. Werde das nochmal probieren.

**Comment by gamer123** [comment]
![Shellgo RSCP__fehler](https://user-images.githubusercontent.com/5868875/167105298-f72e5d93-920d-44cd-aa2b-4451f5689650.jpg)

Das wird Angzeigt. Ich habe den Eintrag EMS_POWER_PV aus dem json gelöscht.

**Comment by gamer123** [comment]
![Shellgo RSCP_fehler_homeassistant](https://user-images.githubusercontent.com/5868875/167106828-90d01f31-aa01-43d8-bfda-0cbfae565fe0.jpg)
Das wird im HA angezeigt. Ist aber klar wenn result leer ist das er dann nichts interpretieren kann.

**Comment by spali** [comment]
Bzgl. EV_POWER, da hat sich im Wiki ein Fehler eingeschlichen. Muss `EMS_REQ_POWER_PV``` heissen. Habe das `e3dc.json` Beispiel im Wiki korrigiert.

Danach sollte die direkte Skript Ausführung in etwa folgendes ausspucken:
```json
{"EMS_POWER_PV":0,"EMS_POWER_BAT":2463,"EMS_POWER_HOME":809,"EMS_POWER_GRID":9,"EMS_POWER_ADD":-3263,"EMS_AUTARKY":99.2228,"EMS_SELF_CONSUMPTION":99.368095,"EMS_BAT_SOC":52,"EMS_GET_POWER_SETTINGS":{"EMS_POWER_LIMITS_USED":true,"EMS_MAX_CHARGE_POWER":3000,"EMS_MAX_DISCHARGE_POWER":2000,"EMS_DISCHARGE_START_POWER":65,"EMS_POWERSAVE_ENABLED":false,"EMS_WEATHER_REGULATED_CHARGE_ENABLED":true,"EMS_WEATHER_FORECAST_MODE":1},"EMS_GET_MANUAL_CHARGE":{"EMS_MANUAL_CHARGE_START_COUNTER":1645508793961,"EMS_MANUAL_CHARGE_ACTIVE":false,"EMS_MANUAL_CHARGE_ENERGY_COUNTER":0.00018570601851851853,"EMS_MANUAL_CHARGE_LASTSTART":"2022-02-22T05:46:33Z"},"INFO_TIME":"2022-05-06T12:04:38.000846Z"}
```
In deinem Fall scheint aber noch ein anderes Problem zu sein.
Die Zeit welche zurück gesendet wird, scheint kein gültiges Jahr zu enthalten...  ich hatte das selber auch mal nach einem Firmware update der E3DC. Konnte es jedoch nicht reproduzieren.... 
Vom Gefühl her verliert die E3DC entweder die Zeit selber, oder die Zeit des letzten manuellen Ladevorgangs wird resetet, und spuckt irgend was komisches als Zeit zurück.

Als Workaround kannst du im `e3dc.json` die Abfragen welche Zeiten zurück spucken auslassen:
```json
[
  "EMS_REQ_BAT_SOC",
  "EMS_REQ_POWER_PV",
  "EMS_REQ_POWER_ADD",
  "EMS_REQ_POWER_BAT",
  "EMS_REQ_POWER_HOME",
  "EMS_REQ_POWER_GRID",
  "EMS_REQ_SELF_CONSUMPTION",
  "EMS_REQ_AUTARKY",
  "EMS_REQ_GET_POWER_SETTINGS"
]
```

Um heraus zu finden welche Zeit krumm ist, kannst du bitte mal folgende beiden Befehle ausführen. Bitte die Ausgabe als Text hier schreiben... sonst wird es extrem mühsam die ausgespuckten Bytes zu analysieren ;)
```
/config/bin/e3dc -config /config/packages/e3dc/.config -debug 6 '["INFO_REQ_TIME"]'
```
```
/config/bin/e3dc -config /config/packages/e3dc/.config -debug 6 '["EMS_REQ_GET_MANUAL_CHARGE"]'
```

**Reply by gamer123** [reply]
Hi @spali,

danke nochmals für deine Hilfe, ich kann es demnächst nochmal testen, aber ich gehe davon aus das es klappt. 
Die erste Zeit INFO_REQ_TIME habe ich schon rausgeworfen bevor ich hier geantwortet habe, weil ich dachte das es an dieser liegt. 

Die zweite Zeit ist jedoch noch im json drin EMS_REQ_GET_MANUAL_CHARGE und ich kann sicher sagen das das letzte mal Manual Charge bei der IBN stattgefunden hat und nach einigen Monaten ein Softwareupdate stattgefunden hat und das es danach kein manuelles Laden mehr gegeben hat. (afaik)

Melde mich nochmal wenn es geklappt hat! 
Evlt. wäre es sinnvoll diese Zeit mit einem Hinweis zu versehen oder erstmal aus dem Beispiel zu entfernen, damit man nicht gleich in so einen intransparenten Fehler läuft. Wenn es am FW Update liegen sollte, dann müsste der Fehler früher oder später bei allen auftreten. 

Danke und super Support und Integration für HA !!!

Grüße 

Julian

**Comment by spali** [comment]
Wenn meine Theorie stimmt, müsste folgender Befehl 
```
/config/bin/e3dc -config /config/packages/e3dc/.config -debug 6 '["EMS_REQ_GET_MANUAL_CHARGE"]'
```
bei dir knallen.

Dazu habe ich nun noch ein potentiellen Fix gebaut:
```
mkdir -p /config/bin
cd /config/bin
curl -L https://github.com/spali/go-rscp/releases/download/v0.1.1-test01/e3dc_0.1.1-test01_Linux_x86_64.tar.gz | tar -xzf - e3dc 
```

Mit dem Fix müsste der Befehl oben wieder funktionieren und das ungültige Datum als `"1970-01-01T00:00:00Z"` ausspucken.

Falls du das so bestätigen kannst, baue ich aus dem Fix noch ein sauberes Release `v0.1.1`.

**Comment by gamer123** [comment]
Hi,
sorry für die späte Antwort. Ich war im Urlaub und habe es dann kurzeitig vergessen.
Ja das hat den Fehler verursacht.
![Screenshot 2022-05-08 193702](https://user-images.githubusercontent.com/5868875/171488076-c5e6f657-fcd2-4d9b-a4e9-21edb0ed8ace.jpg)

Jetzt Funktioniert es. 

Vielen Dank für den Support. 
Ich denke ich werde nochmal fragen haben, wenn ich den E3DC versuchen werden zu steuern. 

Grüße
mujuli
